File: app/src/main/java/gemini/workshop/FunctionCalling.java
/*
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package gemini.workshop;

import dev.langchain4j.model.chat.ChatLanguageModel;
import dev.langchain4j.model.chat.request.json.JsonObjectSchema;
import dev.langchain4j.model.vertexai.VertexAiGeminiChatModel;
import dev.langchain4j.agent.tool.ToolExecutionRequest;
import dev.langchain4j.agent.tool.ToolSpecification;
import dev.langchain4j.data.message.AiMessage;
import dev.langchain4j.data.message.ChatMessage;
import dev.langchain4j.data.message.UserMessage;
import dev.langchain4j.data.message.ToolExecutionResultMessage;
import dev.langchain4j.model.output.Response;

import java.util.List;
import java.util.ArrayList;

public class FunctionCalling {
    public static void main(String[] args) {
        ChatLanguageModel model = VertexAiGeminiChatModel.builder()
            .project(System.getenv("PROJECT_ID"))
            .location(System.getenv("LOCATION"))
            .modelName("gemini-1.5-pro-002")
            .maxOutputTokens(100)
            .build();

        ToolSpecification weatherToolSpec = ToolSpecification.builder()
            .name("getWeather")
            .description("Get the weather forecast for a given location or city")
            .parameters(JsonObjectSchema.builder()
                .addStringProperty(
                    "location",
                    "the location or city to get the weather forecast for")
                .build())
            .build();

        List<ChatMessage> allMessages = new ArrayList<>();

        // 1) Ask about the weather
        UserMessage weatherQuestion = UserMessage.from("What is the weather in Paris?");
        allMessages.add(weatherQuestion);

        // 2) The model replies with a function call request
        Response<AiMessage> messageResponse = model.generate(allMessages, weatherToolSpec);
        ToolExecutionRequest toolExecutionRequest = messageResponse.content().toolExecutionRequests().get(0);
        System.out.println("Tool execution request: " + toolExecutionRequest);
        allMessages.add(messageResponse.content());

        // Here, we would call a real weather forecast service

        // 3) We send back the result of the function call
        ToolExecutionResultMessage toolExecResMsg = ToolExecutionResultMessage.from(toolExecutionRequest,
            "{\"location\":\"Paris\",\"forecast\":\"sunny\", \"temperature\": 20}");
        allMessages.add(toolExecResMsg);

        // 4) The model answers with a sentence describing the weather
        Response<AiMessage> weatherResponse = model.generate(allMessages);
        System.out.println("Answer: " + weatherResponse.content().text());
    }
}

File: python/function_calling.py
import json
import os

from langchain_core.messages import HumanMessage, ToolMessage
from langchain_core.tools import tool
from langchain_google_vertexai import ChatVertexAI


# Reference: https://python.langchain.com/v0.2/docs/how_to/function_calling/

@tool
def get_weather_forecast(location: str) -> str:
    """Get the weather forecast for a given location or city"""
    data = {
        "location": location,
        "forecast": "sunny",
        "temperature": 20
    }
    json_data = json.dumps(data)
    return json_data


if __name__ == "__main__":
    tools = [get_weather_forecast]

    llm = ChatVertexAI(
        project=os.environ["PROJECT_ID"],
        location="us-central1",
        model="gemini-1.5-flash-002"
    )

    llm_with_tools = llm.bind_tools(tools)

    # Ask about the weather
    query = "How's the weather in Paris?"
    print(f"User: {query}")
    messages = [HumanMessage(query)]
    response = llm_with_tools.invoke(messages)
    messages.append(response)

    # The model replies with a function call request
    print(f"Response: {response.tool_calls}")

    for tool_call in response.tool_calls:
        selected_tool = {"get_weather_forecast": get_weather_forecast}[tool_call["name"].lower()]
        tool_output = selected_tool.invoke(tool_call["args"])
        # Send back the result of the function call
        messages.append(ToolMessage(tool_output, tool_call_id=tool_call["id"]))

    # Invoke the model again with function call response
    response = llm_with_tools.invoke(messages)
    print(f"Response: {response.content}")

